File: Algorithms/Flow/Algorithms/Search/DF_Search/DF_Search_cpp/DF_Search.cpp
//
//  DF_Search.cpp
//  Algorithms
//
//  Created by Maksim Ponomarev on 12/4/24.
//

#include "DF_Search.hpp"

#include <iostream>
#include <vector>
#include <unordered_set>

class Graph {
public:
	// Adjacency list representation of the graph
	std::unordered_map<int, std::vector<int>> adjacencyList;

	// Function to add an edge to the graph
	void addEdge(int source, int destination) {
		adjacencyList[source].push_back(destination);
	}

	// DFS function to find path from start to target
	bool dfs(int start, int target, std::unordered_set<int>& visited, std::vector<int>& path) {
		// Mark the node as visited
		visited.insert(start);
		path.push_back(start);

		// Check if we have reached the target
		if (start == target) {
			return true; // Target found
		}

		// Explore neighbors
		for (int neighbor : adjacencyList[start]) {
			if (visited.find(neighbor) == visited.end()) {
				if (dfs(neighbor, target, visited, path)) {
					return true; // Path found in recursive call
				}
			}
		}

		// Backtrack if no path is found
		path.pop_back();
		return false; // No path found
	}

	// Function to find a path from start to target
	std::vector<int> findPath(int start, int target) {
		std::unordered_set<int> visited;
		std::vector<int> path;

		if (dfs(start, target, visited, path)) {
			return path; // Return the found path
		} else {
			return {}; // Return an empty vector if no path found
		}
	}
};

// Example usage
int main() {
	Graph graph;
	graph.addEdge(0, 1);
	graph.addEdge(0, 2);
	graph.addEdge(1, 3);
	graph.addEdge(2, 3);
	graph.addEdge(3, 4);

	int start = 0, target = 4;
	std::vector<int> path = graph.findPath(start, target);

	if (!path.empty()) {
		std::cout << "Path found: ";
		for (int node : path) {
			std::cout << node << " ";
		}
		std::cout << std::endl;
	} else {
		std::cout << "No path found." << std::endl;
	}

	return 0;
}

File: Algorithms/Flow/Algorithms/Search/DF_Search/DFSearch.swift
//
//  DFSearch.swift
//  Algorithms
//
//  Created by Maksim Ponomarev on 12/4/24.
//
/*
 Explanation:

 Graph Representation:
 The graph is represented as an adjacency list using a dictionary.
 
 DFS Function:
 The dfs function recursively explores each node. It uses a visited set to keep track of visited nodes and a path array to record the current path being explored.

 Backtracking:
 If a path does not lead to the target, it backtracks by removing the last node from the path.
 Finding the Path: The findPath method initializes the search.

 Note:
 This implementation does not guarantee the shortest path due to the nature of DFS. For finding the shortest path in an unweighted graph, it is more common to use BFS.
 
 Depth-First Search (DFS) is typically used to explore all possible paths in a graph. While it’s more common to use Breadth-First Search (BFS) for finding the shortest path in an unweighted graph, you can combine DFS with backtracking to find paths. A typical DFS implementation does not guarantee the shortest path, DFS to explore paths.
 */

import Foundation

class Graph {
	var adjacencyList: [Int: [Int]] = [:]

	func addEdge(from source: Int, to destination: Int) {
		adjacencyList[source, default: []].append(destination)
	}
	
	func dfs(start: Int, target: Int, visited: inout Set<Int>, path: inout [Int]) -> [Int]? {
		// Mark the node as visited
		visited.insert(start)
		path.append(start)

		// Check if we have reached the target
		if start == target {
			return path
		}

		// Explore neighbors
		for neighbor in adjacencyList[start, default: []] {
			if !visited.contains(neighbor) {
				if let resultPath = dfs(start: neighbor, target: target, visited: &visited, path: &path) {
					return resultPath
				}
			}
		}

		// Backtrack
		path.removeLast()
		return nil
	}

	func findPath(start: Int, target: Int) -> [Int]? {
		var visited = Set<Int>()
		var path = [Int]()
		return dfs(start: start, target: target, visited: &visited, path: &path)
	}
}

/*
// Example usage
let graph = Graph()
graph.addEdge(from: 0, to: 1)
graph.addEdge(from: 0, to: 2)
graph.addEdge(from: 1, to: 3)
graph.addEdge(from: 2, to: 3)
graph.addEdge(from: 3, to: 4)

if let path = graph.findPath(start: 0, target: 4) {
	print("Path found: \(path)")
} else {
	print("No path found.")
}
*/
